### PR TITLE
Start working on type ascriptions

### PR DESCRIPTION
**Pull request by vendethiel**
I would like everyone's input on this: the syntax, what we're compiling to, etc.
The idea is to compile down to TS or flow-compatible stuff

### CODE REVIEW

**Comment by heavyk** [comment]
could you give an example? I'm not sure I understand

**Comment by robotlolita** [comment]
So, basically:

```ls
(a @: Type, b @: Type) -> ...
```

Gets compiled to:

```js
(function(a : Type, b : Type){ ... })
```

Is the idea to support things like TypeScript or Flow?

**Comment by vendethiel** [comment]
Yes, indeed.

**Comment by robotlolita** [comment]
Ah. I'm not particularly fond of the `@:` symbol, but I also can't think of any other that is available and denotes a typing relationship well :(

**Comment by vendethiel** [comment]
Right? `:` and `::` both doesn't work, so... I went for the one I could think of.

**Comment by heavyk** [comment]
oh, I see now. yea, I like the idea of compiling to TS/flow.

could this work?
```ls
fn = (:a Type, :b Number) ->
```

**Comment by vendethiel** [comment]
It'd make the lexer a bit more complex, which I'm not too fond of. But I'm open to opinions.

**Comment by robotlolita** [comment]
An alternative might be using the same symbol Clojure uses to declare types. It's reusing a symbol, but I don't think that's an issue for argument names:

```ls
foo = (a ^Type, b ^Another-Type) ^ReturnType -> ...
foo = (a ^Type = ^clone) -> ...
```

**Comment by vendethiel** [comment]
Yeah, but it's currently parsed as an infix MATH (power) :/

**Comment by vendethiel** [comment]
At least, the fact our unary clone got changed to `^^` helps (because we allow unary ops there)

**Comment by waynedpj** [comment]
> Right? : and :: both doesn't work, so... I went for the one I could think of.

is there any interest in making some breaking changes to free up operators like `:` and `::` that many already recognize as defining types, particularly `:` used in TypeScript+Flow?  IMHO the type ascriptions would take precedence over most other uses of operators.

also, even the Livescript homepage seemingly uses `::` for type ascriptions (in the 'Programmatic API' section):
`LiveScript.compile(code :: String, options :: Object?) -> String`

finally, there already seems to be some discussion on removing duplicated/etc. functionality:
https://github.com/gkz/LiveScript/issues/219
https://github.com/gkz/LiveScript/issues/721

peace

**Comment by vendethiel** [comment]
We can change the docs, that's very minor. It currently just mimics Haskell's syntax :-).

You're free to ping those issues to try and get some tractions ;-).

**Comment by gkz** [comment]
(x :: Number) ->

Seems to be a syntax error - maybe we could just support the syntax for function definitions?

**Comment by vendethiel** [comment]
Problem is just the lexer's complexity. We need to resort to per-nest flag (or something) to parse correctly `(a :: Foo = -> ::) ->`, say

**Comment by vendethiel** [comment]
Urgh, https://github.com/gkz/LiveScript/blob/master/test/cs-classes.ls#L289 doesn't parse correctly. Obviously. I guess I need to change the regexp to `@:(?!:)`...

**Comment by heavyk** [comment]
ok, if you don't like the atom syntax, what about c++ syntax?

```ls
lala = (a<Type>, b<Number> = 3) -> ...
```

**Comment by rightfold** [comment]
`a isa number`

**Comment by igl** [comment]
Arrow functions are fine and dandy but what aboot  braceless function declarations? 
```
function add : number a : number, b : number => a + b
function add @ number a @ number, b@number => a + b
function add @: number a @:number, b @:number => a + b
function add~number a~number, b~number => a + b
function add^number a^number, b^number => a + b
function add<number> a<number>, b<number> => a + b
```

C++ style looks best to me.

What do you guys think about type contracts like in contracts.coffee or elm?

**Comment by vendethiel** [comment]
I really don't like parenless function definitions, but anyway, all the signatures will look the same.

**Comment by unclechu** [comment]
I vote for `::` it's easy to catch by eyes for me and I think this:
```text
fn = (a :: Type, b :: Number) ->
```
is more obvious than that:
```text
fn = (:a Type, :b Number) ->
```

**Comment by coachshea** [comment]
agreed

**Comment by igl** [comment]
Unfortunately `::` compiles to `prototype` in LS.
There is also a es7 proposal using `::` it's only a proposal though.

**Comment by AriaMinaei** [comment]
How about declaring the function type in a separate line?

```livescript
add :: (Number, Number) -> Number
add = (a, b) ->
  a + b
```

I think the other syntax (defining the type in the same line as declaring the function) creates too much visual clutter, and it just doesn't go with LiveScript's terse syntax, but that might be a matter of taste.

**Comment by unclechu** [comment]
@AriaMinaei may be as alternative, but if you have a lot of arguments, inline declaration will be easier to read.
```livescript
add :: (
  number
  number
  string
  boolean
  Object
  Date
  string[]
) -> number
add = (
  foo
  bar
  test
  test2
  abc
  def
  zyx
) ->
  a + b
```
In this case I can't say immediately what type does have `test2`.
And in case with inline type declaration we keep it short and readable:
```livescript
add = (
  :foo   number
  :bar   number
  :test  string
  :test2 boolean
  :abc   Object
  :def   Date
  :zyx   string[]
): number ->
  a + b

**Comment by AriaMinaei** [comment]
@unclechu Correct, but how many functions *would* have that many arguments? A function invocation with more than five or six arguments would be a bit unreadable, unless the arguments are named (through a config object).

Also, writing type annotations in the same line as function declaration makes even simple functions like [this](https://github.com/AriaMinaei/shapely/blob/master/src/shapely/validators/ExactValueValidator.js#L18) overly verbose. I'm not even sure if that *is* a matter of taste.

**Comment by focused** [comment]
I love this one style of type declaration:
```
add :: (Number, Number) -> Number
add = (a, b) ->
    a + b
```
@unclechu I think we should not pass more than 4 or 5 arguments, in this case we should use data structures (Array, Object). Look:
```
count-leaves :: (Tree, BranchSide) -> Number
count-leaves = (t, s) ->
  ...
```

**Comment by isiahmeadows** [comment]
@vendethiel Status?

**Comment by vendethiel** [comment]
@isiahmeadows still undecided on which token. Overloading `::` seems risky (and hard to parse in cases like `(a = -> x::) ->`, say).

**Comment by thelambdaparty** [comment]
Maybe [JavaScript types in TypeScript](https://github.com/Microsoft/TypeScript/issues/4790) ?
Uses JSDoc for type information.

**Comment by isiahmeadows** [comment]
@thelambdaparty I really dislike JSDoc type annotations. They are almost a joke in their verbosity.

**Comment by rightfold** [comment]
> They are almost a joke

As is TypeScript.

**Comment by isiahmeadows** [comment]
@rightfold To be honest, *Java's* type system are better than JSDoc annotations in my opinion. And I really don't like Java. The only thing they're worth using for are for external API documentation, and even that's disputable.

I wasn't comparing them to powerful type systems like Haskell's.

**Comment by rightfold** [comment]
Who said anything about Java?

**Comment by isiahmeadows** [comment]
I was drawing a comparison to clarify what I meant. Basically, I feel JSDoc
types would suck more than the type system of a really bad language.
Especially for a concise language.

**Comment by kress95** [comment]
So many options...

    (a & Type, b & Type) -> ...
    (a  @: Type, b @: Type) -> ...
    (a ::{Type}, b ::{Type}) -> ...
    (a ::<Type>, b ::<Type>) -> ...
    (a &<Type>, b &<Type>) -> ...
    (a <:Type:>, b <:Type:>) -> ...
    (a {:Type:}, b {:Type:}) -> ...
    (a {.Type.}, b {.Type.}) -> ...

(I really like the syntax of the last two, reminds me of nim's pragmas)

**Comment by isiahmeadows** [comment]
I'm okay with any of those except the ones with ampersands (I would
consistently misread them as type union, which TypeScript isn't the only
one using that syntax).

**Comment by kay-is** [comment]
Do they need to be non-valid js identifiers (like : or ^ etc.)? 

If they can be anything, how about `§` ?

`§` is a bit associated with laws (which types essentially are)

    my-func = (a §Number, b §String) §Boolean -> 
    my-func = (a §Number, b §String) -> §Boolean 

    my-func = §Boolean (§Number a, §String b)->

**Comment by isiahmeadows** [comment]
I would prefer ASCII symbols to denote types, BTW. And preferably, yes,
they should. I already run into issues occasionally with `is` and `_`.

**Comment by heavyk** [comment]
I have the § key on my keyboard and I would have suggested it also but the us kbd does not have the key. I always thought it was ascii, but upon looking - alas it is not - I guess it was one of the iso standards.

I'm also ok with whatever cept for ampersands. I suppose the clearest to my eyes is the `a @: Type` one

**Comment by focused** [comment]
I have no such key, and many of us

**Comment by heavyk** [comment]
sorry, that wasn't the point.

> I suppose the clearest to my eyes is the `a @: Type` one

**Comment by rhendric** [comment]
I know this is a closed PR, but since @dk00 is still linking to it, it seems useful as a repository for bikeshed suggestions. Here's mine:

`(a (Number), b (String)) (Boolean) -> ...`
`x (foo: String || Number) = get-foo!`

Advantage of this syntax over most of the above: it already parses into an AST, and definitely won't conflict with any valid LS since you can't assign to a function call.

**Comment by vendethiel** [comment]
I'm not sure what that 2nd line describes?

**Comment by rhendric** [comment]
Ascriptions for variable declaration.

**Comment by vendethiel** [comment]
is `x` or `foo` the variable?

**Comment by rhendric** [comment]
`x` is a variable, with a type that asserts it has a property named `foo` with value that is a `String` or a `Number`. As elsewhere in LS, `foo: String || Number` is shorthand for `{ foo: String || Number }`.

**Comment by vendethiel** [comment]
Isn't this ambiguous with `x((foo: String || Number) = get-foo!)` (declare a variable inside a funcall)?

**Comment by rhendric** [comment]
Gah, yes, you're right! Righto, that's what I get for tossing off half-baked thoughts in passing.

**Comment by vendethiel** [comment]
It might be better, at least at first, to only allow type ascription when using declarators (`var`/`const`) (maybe in `let`s as well).

**Comment by ozra** [comment]
Another soft ball in the air: `foo = (x ^Number, y ^String) ^String -> "#{x}, #{y}"`.

Too bad @kay-is suggestion of `§` isn't us-keymap and us-charset friendly, I considered that one in my pet lang before too :-/

The Haskellish way posed earlier in the issue is a good bet too, and when non-clashing symbology can be figured out - _both_ the inline- and pre- typedecl would be awesome imo.

**Comment by vendethiel** [comment]
Well, there's always the solution of writing the type signature before the object itself, and not inline. That'd remove a good chunk of ambiguities.

**Comment by danielo515** [comment]
What is the status of this ?

**Comment by rhendric** [comment]
Exactly what it looks like. There's the occasional idea being tossed around in #999, if you feel like helping out.

**Comment by danielo515** [comment]
I'll take a look at that issue.
Thanks
